tests/Integration/Behaviour/Features/Context/Domain/CustomerGroupFeatureContext.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+namespace Tests\Integration\Behaviour\Features\Context\Domain;
+
+use Behat\Gherkin\Node\TableNode;
+use Exception;
+use PrestaShop\PrestaShop\Core\CommandBus\CommandBusInterface;
+use PrestaShop\PrestaShop\Core\Domain\Customer\Group\Command\AddCustomerGroupCommand;
+use PrestaShop\PrestaShop\Core\Domain\Customer\Group\Query\GetCustomerGroupForEditing;
+use PrestaShop\PrestaShop\Core\Domain\Customer\Group\QueryResult\EditableCustomerGroup;
+use PrestaShop\PrestaShop\Core\Domain\Customer\Group\ValueObject\GroupId;
+use PrestaShop\PrestaShop\Core\Domain\Product\QueryResult\ProductForEditing;
+use Tests\Integration\Behaviour\Features\Context\CommonFeatureContext;
+use Tests\Integration\Behaviour\Features\Context\Domain\Product\UpdateTagsFeatureContext;
+use Tests\Integration\Behaviour\Features\Context\Util\DataComparator;
+
+class CustomerGroupFeatureContext extends AbstractDomainFeatureContext
+{
+    /**
+     * "When" steps perform actions, and some of them store the latest result
+     * in this variable so that "Then" action can check its content
+     *
+     * @var mixed
+     */
+    protected $latestResult;
+
+    /**
+     * Registry to keep track of created/edited customer groups using references
+     *
+     * @var int[]
+     */
+    protected $registry = [];
+
+    /**
+     * @When /^I create a Customer Group "(.+)" with the following details:$/
+     *
+     * @param string $customerGroupReference
+     * @param TableNode $table
+     *
+     * @throws Exception
+     */
+    public function createCustomerUsingCommand(string $customerGroupReference, TableNode $table)
+    {
+        $data = $this->localizeByRows($table);
+        $commandBus = $this->getCommandBus();
+
+        $command = new AddCustomerGroupCommand(
+            $data['name'],
+            $data['reduction'],
+            $data['priceDisplayMethod'],
+            $data['showPrice']
+        );
+
+        /** @var GroupId $id */
+        $id = $commandBus->handle($command);
+        $this->latestResult = $id->getValue();
+        $this->registry[$customerGroupReference] = $id->getValue();
+    }
+
+    /**
+     * @When /^I query Customer Group "(.+)" I should get a Customer Group with properties:$/
+     */
+    public function assertQueryCustomerProperties($customerGroupReference, TableNode $table)
+    {
+        $expectedData = $table->getRowsHash();
+
+        $queryBus = $this->getQueryBus();
+        /** @var EditableCustomerGroup $result */
+        $result = $queryBus->handle(new GetCustomerGroupForEditing($this->registry[$customerGroupReference]));
+
+        $serializer = CommonFeatureContext::getContainer()->get('serializer');
+        $realData = $serializer->normalize($result);
+
+        DataComparator::assertDataSetsAreIdentical($expectedData, $realData);
```

Comment:
interesting.
Not sure if this will work when you need to convert some of the values to DecimalNumber or some other classes or allow partial assertions. You might need to implement manual assertion then.

If you still want to always assert full EditableCustomerGroup, I suggest you use `Transform` instead of this serializer and then compare two objects with assertEqual. See transformation example in
DetailsAssertionFeatureContext::transformDetails.

What it will make is convert your TableNode to your desired object (in this case EditableCustomerGroup), so this assertion method would end up looking smth like:
```
public function assertQueryCustomerProperties($customerGroupReference, EditableCustomerGroup $expectedGroup): void
{
    Assert::assertEquals($this->getCustomerGroupForEditing(), $expectedGroup);
}
```

Reply:
Yep interesting approach, but I agree it can get complicated with the DecimalNumber fields that you need to use in QueryResult (we replace piece by piece all the usage of built-in floats because of their lack of precision)

Reply:
I went with transformDetails and works fine, also makes it easier to test localized values, double win 👌